crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -204,8 +205,11 @@ impl CheckpointExecutor {
 
                 }
                 // Check for newly synced checkpoints from StateSync.
-                received = self.mailbox.recv() => match received {
-                    Ok(checkpoint) => {
+                received = timeout(Duration::from_millis(SCHEDULING_EVENT_FUTURE_TIMEOUT_MS), self.mailbox.recv()) => match received {
```

Comment:
I wonder if it is more readable if we structure the code into

```
tokio::select! {
    _ = pending.next() => {}
    _ = self.mailbox.recv() => {}
    _ = sleep(timeout_duration) => {}
}
```

instead of only adding a timeout in the second branch and treat the timeout as an error.

Reply:
Yeah I had originally considered doing that. The problem is that `tokio::time::sleep` doesn't implement unpin so I would have to manually re-allocate and pin the sleep future if using in a `tokio::select`. See [here](https://docs.rs/tokio/latest/tokio/time/struct.Sleep.html). Just seemed more straightforward to do what I'm doing here. Semantically it makes perfect sense to me - we wait for the first of two options to occur: either a newly completed checkpoint execution or a newly synced checkpoint. If we're stuck in execution too long, we have other logging in checkpoint executor, so in a sense, the only thing we need to be timing here is receiving a newly synced checkpoint.

Reply:
Sounds good! Optionally, maybe worth documenting here why we add a timeout.